src/lkl/setup.c

```diff
@@ -1211,45 +1252,43 @@ static void* lkl_termination_thread(void* args)
             lkl_strerror(ret));
     }
 
-    // Unmount disks
+    // Unmount mounts
     long res;
-    for (int i = num_disks - 1; i >= 0; --i)
+    for (int i = sgxlkl_enclave_state.num_disk_state - 1; i > 0; --i)
     {
-        if (!disks[i].mounted)
+        if (!sgxlkl_enclave_state.disk_state[i].mounted)
             continue;
 
-        SGXLKL_VERBOSE("calling lkl_umount_timeout(%s)\n", disks[i].mnt);
-        res = lkl_umount_timeout(disks[i].mnt, 0, UMOUNT_DISK_TIMEOUT);
+        sgxlkl_enclave_mount_config_t* disk_i = &cfg->mounts[i];
+        SGXLKL_VERBOSE("calling lkl_umount_timeout(%s)\n", disk_i->destination);
+        res = lkl_umount_timeout(disk_i->destination, 0, UMOUNT_DISK_TIMEOUT);
         if (res < 0)
         {
             sgxlkl_warn(
                 "Could not unmount disk %d, %s\n", i, lkl_strerror(res));
         }
 
-        // Root disk, no need to remove mount point ("/").
-        if (i == 0)
-            continue;
-
-        // Read-only root disk, cannot remove pre-existing mount points for
-        // secondary disks.
-        if (disks[0].ro)
-            continue;
-
         // Not really necessary for mounts in /mnt since /mnt is
         // mounted as tmpfs itself, but it is also possible to mount
         // secondary images at any place in the root file system,
         // including persistent storage, if the root file system is
         // writeable. For simplicity, remove all mount points here.
-        res = lkl_sys_rmdir(disks[i].mnt);
+        res = lkl_sys_rmdir(disk_i->destination);
         if (res < 0)
         {
             sgxlkl_warn(
                 "Could not remove mount point %s\n",
-                disks[i].mnt,
+                disk_i->destination,
                 lkl_strerror(res));
         }
     }
 
+    // Unmount root
+    SGXLKL_VERBOSE("calling lkl_umount_timeout(/)\n");
+    res = lkl_umount_timeout("/", 0, UMOUNT_DISK_TIMEOUT);
+    if (res < 0)
+        sgxlkl_warn("Could not unmount root disk, %s\n", lkl_strerror(res));
```

Comment:
Ah, you are unmounting the root disk last?

Reply:
Yes, but only because the root disk is now not included in the mounts anymore (OCI structure). I can move it elsewhere but I think this should be fine, no?

Reply:
Looks fine to me.